File: dotnet/Servirtium.Vcr.Integration.TodoBackend/TodoBackendPlaybackIT.cs
using Servirtium.Vcr;
using Xunit;

namespace Servirtium.Vcr.Integration.TodoBackend;

/// <summary>
/// TodoBackend browser integration test — PLAYBACK phase (the CI artifact).
///
/// <para>Replays the committed CRUD tape through a Servirtium VCR and runs the
/// real TodoBackend Mocha spec against it in real headless Chrome (.NET's own
/// Selenium). No SUT, no network — the whole CRUD conversation comes off the
/// tape. Mirrors the Python <c>playback_test.py</c>.</para>
///
/// <para>Run via the leaf <c>integration/todobackend/.dotnet_playback.ae</c>, or
/// directly with the working dir at <c>dotnet/</c>:</para>
/// <code>
///   SERVIRTIUM_VCR_LIB=../core/native/libservirtium_vcr.so \
///     dotnet test Servirtium.Vcr.Integration.TodoBackend --nologo \
///     --filter FullyQualifiedName!~TodoBackendRecord
/// </code>
/// </summary>
public class TodoBackendPlaybackIT
{
    [Fact]
    public void MochaSuitePassesOffTheTape()
    {
        using var vcr = Vcr.Playback(TodoBackendBrowser.Tape)
            .StaticContent("/suite", TodoBackendBrowser.SuiteDir)
            .Untaped("/favicon.ico")
            .Port(TodoBackendBrowser.VcrPort)
            .Start();

        var r = TodoBackendBrowser.RunSuite(vcr.BaseUrl);
        System.Console.WriteLine($"mocha (playback): {r.Passes} passed, {r.Failures} failed");
        foreach (string m in r.FailMessages)
        {
            System.Console.WriteLine("  FAIL: " + m);
        }

        Assert.True(r.Failures == 0, "mocha reported failures: " + string.Join("; ", r.FailMessages));
        Assert.True(r.Passes > 0, "mocha reported no passes");
    }
}

File: java/src/test/java/com/paulhammant/servirtium/vcr/todobackend/TodoBackendPlaybackIT.java
package com.paulhammant.servirtium.vcr.todobackend;

import com.paulhammant.servirtium.vcr.Vcr;
import com.paulhammant.servirtium.vcr.VcrServer;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * TodoBackend browser integration test — PLAYBACK phase (the CI artifact).
 *
 * <p>Replays the committed CRUD tape through a Servirtium VCR and runs the real
 * TodoBackend Mocha spec against it in real headless Chrome (Java's own
 * Selenium). No SUT, no network — the whole CRUD conversation comes off the
 * tape. Mirrors the Python {@code playback_test.py}.
 *
 * <p>Run via the leaf {@code integration/todobackend/.java_playback.ae}, or
 * directly with the working dir at {@code java/}:
 * <pre>{@code
 *   SERVIRTIUM_VCR_LIB=../core/native/libservirtium_vcr.so \
 *     mvn -q test -Dtest=TodoBackendPlaybackIT
 * }</pre>
 */
class TodoBackendPlaybackIT {

    @Test
    void mochaSuitePassesOffTheTape() {
        try (VcrServer vcr = Vcr.playback(TodoBackendBrowser.TAPE.toString())
                .staticContent("/suite", TodoBackendBrowser.SUITE_DIR.toString())
                .untaped("/favicon.ico")
                .port(TodoBackendBrowser.VCR_PORT)
                .start()) {

            TodoBackendBrowser.Result r = TodoBackendBrowser.runSuite(vcr.baseUrl());
            System.out.printf("mocha (playback): %d passed, %d failed%n", r.passes(), r.failures());
            for (String m : r.failMessages()) {
                System.out.println("  FAIL: " + m);
            }

            assertEquals(0, r.failures(), "mocha reported failures: " + r.failMessages());
            assertTrue(r.passes() > 0, "mocha reported no passes");
        }
    }
}
